### PR TITLE
bootstrap repository

### PR DESCRIPTION
**Pull request by cjp256**
- Sphinx documentation and config.

- Read the docs config.

- Codespell config for code spelling.

- Pylint config for linting.

- Flake8 config for linting.

- Pycodestyle config for linting python code.

- Pydocstyle config for linting python docs.

- Pyright config for static analysis.

- Mypy configuration for static analysis.

- Isort config (pyproject.toml) for organizing imports.

- Black config for formatting.

- Craft providers packaging.

- Ported cla-check from snapcore with some fixes to pass linting.

- Initial package layout for tests (integration and unit).

- Initial package layout for craft_providers main package.

- Starter tox configuration, though unused.

- Initial README.md.

- Makefile to expose all entry points and targets.

- Github Actions / Workflows / Configuration to cover all tests.

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python
+#
+# Copyright (C) 2021 Canonical Ltd
```
Comment (sergiusens) [inline setup.py:3]:
```suggestion
# Copyright 2021 Canonical Ltd.
```

File: setup.cfg [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+extend-ignore = E203, E501, W503
+
+[mypy]
+python_version = 3.8
+
+[pycodestyle]
+ignore = E402, E501, W503
+
+[pydocstyle]
+ignore = D105, D107, D203, D204, D213, D215
```
Comment (sergiusens) [inline setup.cfg:37]: Maybe add a comment on what each these exceptions are (for all ignores)

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    long_description=readme,
+    include_package_data=True,
+    keywords="craft_providers",
+    name="craft-providers",
+    package_data={"craft_providers": ["py.typed"]},
+    packages=find_packages(include=["craft_providers", "craft_providers.*"]),
+    setup_requires=setup_requirements,
+    test_suite="tests",
+    tests_require=test_requirements,
+    url="https://github.com/cjp256/craft_providers",
```
Comment (sergiusens) [inline setup.py:65]:
```suggestion
    url="https://github.com/canonical/craft_providers",
```

File: tests/unit/test_todo.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# Copyright (C) 2021 Canonical Ltd
```
Comment (sergiusens) [inline tests/unit/test_todo.py:1]:
```suggestion
# Copyright 2021 Canonical Ltd.
```

Comment (sergiusens) [reply tests/unit/test_todo.py:1]: and so on and so forth

**Review by sergiusens — CHANGES_REQUESTED** [review]

File: setup.cfg (hunk 2 above)
Comment (cjp256) [reply setup.cfg:37]: I will do that.  FWIW most of these are already ignored in the default conventions:
http://www.pydocstyle.org/en/stable/error_codes.html#default-conventions

File: tests/unit/test_todo.py (hunk 4 above)
Comment (cjp256) [reply tests/unit/test_todo.py]: fixed

File: setup.cfg (hunk 2 above)
Comment (cjp256) [reply setup.cfg]: added explanations, also removed/regenerated configs for other linters (will add explanations & disable as required)

File: setup.py (hunk 1 above)
Comment (cjp256) [reply setup.py]: fixed

File: .github/workflows/tests.yaml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+          make test-black
+      - name: Run codespell
+        run: |
+          source ${HOME}/.venv/bin/activate
+          make test-codespell
+      - name: Run flake8
+        run: |
+          source ${HOME}/.venv/bin/activate
+          make test-flake8
+      - name: Run mypy
```
Comment (cmatsuoka) [inline .github/workflows/tests.yaml:43]: Any special reason for not running isort in the GitHub test workflow?

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    long_description=readme,
+    include_package_data=True,
+    keywords="craft_providers",
+    name="craft-providers",
+    package_data={"craft_providers": ["py.typed"]},
+    packages=find_packages(include=["craft_providers", "craft_providers.*"]),
+    setup_requires=setup_requirements,
+    test_suite="tests",
+    tests_require=test_requirements,
+    url="https://github.com/canonical/craft_providers",
```
Comment (sergiusens) [inline setup.py:65]:
```suggestion
    url="https://github.com/canonical/craft-providers",
```

File: .github/workflows/tests.yaml (hunk 5 above)
Comment (cjp256) [reply .github/workflows/tests.yaml:43]: Must have missed it, will add :D

Comment (cjp256) [reply .github/workflows/tests.yaml:47]: Good eye btw, fixed!

**Review by sergiusens — APPROVED** [review]

**Pull request merged**